cmd/simplified-pipeline-ctrl/reconciler.go

```diff
@@ -0,0 +1,153 @@
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	defer cancel()
+	var pj v1.ProwJob
+	if err := r.pjclientset.Get(ctx, req.NamespacedName, &pj); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+
+		return fmt.Errorf("failed to get prowjob %s: %w", req.String(), err)
+	}
+
+	if pj.Spec.Refs == nil || pj.Spec.Type != v1.PresubmitJob {
+		return nil
+	}
+
+	presubmits := r.configDataProvider.GetPresubmits(pj.Spec.Refs.Org + "/" + pj.Spec.Refs.Repo)
+	if len(presubmits.protected) == 0 && len(presubmits.alwaysRequired) == 0 && len(presubmits.conditionallyRequired) == 0 {
+		return nil
+	}
+
+	baseSha, ok := r.baseShas[composeKey(pj.Spec.Refs)]
+	if !ok || (ok && baseSha != pj.Spec.Refs.BaseSHA) {
+		log.Debug("processing next key")
+		delete(r.baseShas, composeKey(pj.Spec.Refs)) // delete the old baseSha
+		status, err := r.reportSuccessOnPR(ctx, &pj, presubmits)
+		if err != nil {
+			return err
+		}
+		if status {
+			if err := r.ghc.CreateCommentWithContext(ctx, pj.Spec.Refs.Org, pj.Spec.Refs.Repo, pj.Spec.Refs.Pulls[0].Number, "/test remaining-required"); err != nil {
+				return fmt.Errorf("error creating comment: %w", err)
+			} else {
+				r.baseShas[composeKey(pj.Spec.Refs)] = pj.Spec.Refs.BaseSHA
+			}
+		}
+	}
+
+	return nil
+}
+
+func composeKey(refs *v1.Refs) string {
+	return fmt.Sprintf("%d:%s/%s@%s", refs.Pulls[0].Number, refs.Org, refs.Repo, refs.BaseRef)
+}
+
+func (r *reconciler) reportSuccessOnPR(ctx context.Context, pj *v1.ProwJob, presubmits presubmitTests) (bool, error) {
+	if pj == nil || pj.Spec.Refs == nil || len(pj.Spec.Refs.Pulls) != 1 {
+		return false, nil
+	}
+	selector := map[string]string{}
+	for _, l := range []string{kube.OrgLabel, kube.RepoLabel, kube.PullLabel} {
+		selector[l] = pj.ObjectMeta.Labels[l]
+	}
+	var pjs v1.ProwJobList
+	if err := r.lister.List(ctx, &pjs, ctrlruntimeclient.MatchingLabels(selector)); err != nil {
+		return false, fmt.Errorf("cannot list prowjob using selector %v", selector)
+	}
+
+	latestBatch := make(map[string]v1.ProwJob)
+	for _, pjob := range pjs.Items {
+		if !pjob.Complete() {
+			return false, nil
+		}
+		if !pjob.Spec.Report {
+			continue
+		}
+		if existing, ok := latestBatch[pjob.Spec.Job]; !ok {
+			latestBatch[pjob.Spec.Job] = pjob
+		} else if pjob.CreationTimestamp.After(existing.CreationTimestamp.Time) {
+			latestBatch[pjob.Spec.Job] = pjob
+		}
+	}
+	// if any of the presubmits is in the latest batch, pipeline was already triggered
+	// unless someone manually triggered it
+	for _, presubmit := range presubmits.protected {
+		if _, ok := latestBatch[presubmit]; ok {
+			return true, nil
+		}
+	}
+	for _, presubmit := range presubmits.alwaysRequired {
+		if pjob, ok := latestBatch[presubmit]; ok && pjob.Status.State != v1.SuccessState {
+			return false, nil
+		}
+	}
+	for _, presubmit := range presubmits.conditionallyRequired {
+		if pjob, ok := latestBatch[presubmit]; ok && pjob.Status.State != v1.SuccessState {
+			return false, nil
+		}
+	}
```

Comment:
This distinction between always/conditionally required doesn't seem to be used anywhere?

Reply:
If we merge them, naming them something to the effect of "first stage" and "second stage" would make it easier to understand the categorization here and in the provider.

Reply:
There was a distinction but I removed it to clear logs. Thanks for the catch.